libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -1719,3 +1701,191 @@ Type const& IRGeneratorForStatements::type(Expression const& _expression)
+			define(
+				m_context.addLocalVariable(*varDecl),
+				IRVariable(*_returnInfo.functionCall).tupleComponent(i++)
+			);
+		}
+	}
+	else
+	{
+		define(
+			m_context.addLocalVariable(*_successClause.parameters()->parameters().front()),
+			IRVariable(*_returnInfo.functionCall)
+		);
+	}
+}
+
+void IRGeneratorForStatements::handleCatch(TryStatement const& _tryStatement)
+{
+	if (_tryStatement.structuredClause())
+		handleCatchStructured(*_tryStatement.structuredClause(), _tryStatement.fallbackClause());
+	else if (_tryStatement.fallbackClause())
+		handleCatchFallback(*_tryStatement.fallbackClause());
+	else
+		rethrow();
+}
+
+void IRGeneratorForStatements::handleCatchStructured(TryCatchClause const& _structured, TryCatchClause const* _fallback)
+{
+	solAssert(
+		_structured.parameters() &&
+		_structured.parameters()->parameters().size() == 1 &&
+		_structured.parameters()->parameters().front() &&
+		*_structured.parameters()->parameters().front()->annotation().type == *TypeProvider::stringMemory(),
+		""
+	);
+	solAssert(m_context.evmVersion().supportsReturndata(), "");
+
+	// Try to decode the error message.
+	// If this fails, leaves 0 on the stack, otherwise the pointer to the data string.
+	string const dataVariable = m_context.newYulVariable();
+
+	m_code << "let " << dataVariable << " := " << m_utils.tryDecodeErrorMessageFunction() << "()\n";
+	m_code << "switch iszero(" << dataVariable << ") \n";
+	m_code << "case 0 { // decoding success\n";
+	if (_structured.parameters())
+	{
+		solAssert(_structured.parameters()->parameters().size() == 1, "");
+		IRVariable const& var = m_context.addLocalVariable(*_structured.parameters()->parameters().front());
+		declare(var);
+		m_code << var.commaSeparatedList() << " := " << dataVariable << "\n";
+	}
+	_structured.accept(*this);
+	m_code << "}\n";
+	m_code << "default { // decoding failure\n";
+	if (_fallback)
+		handleCatchFallback(*_fallback);
+	else
+		rethrow();
+	m_code << "}\n";
+}
+
+void IRGeneratorForStatements::handleCatchFallback(TryCatchClause const& _fallback)
+{
+	if (_fallback.parameters())
+	{
+		solAssert(m_context.evmVersion().supportsReturndata(), "");
+		solAssert(
+			_fallback.parameters()->parameters().size() == 1 &&
+			_fallback.parameters()->parameters().front() &&
+			*_fallback.parameters()->parameters().front()->annotation().type == *TypeProvider::bytesMemory(),
+			""
+		);
+		VariableDeclaration const& paramDecl = *_fallback.parameters()->parameters().front();
+		IRVariable const& parameterVariable = m_context.addLocalVariable(paramDecl);
+		string const parameterVariableName = parameterVariable.commaSeparatedList();
+		if (!m_context.evmVersion().supportsReturndata())
+			m_code << "let " << parameterVariableName << " := " << CompilerUtils::zeroPointer;
+		else
+			m_code << Whiskers(R"(
+				let <v> := returndatasize()
+				switch <v>
+				case 0 {
+					<v> := 0x60
+				}
+				default {
+					// TODO: allocate via allocateMemoryArrayFunction(...)
```

Comment:
This still needs to be done, right?

Reply:
Actually yes, this was a proposal from @ekpyron  ; I just tried right now, but seems like i was running into a NotImplemented-alike exception. Skipping for now.